Source/Particles/Resampling/GridBasedMerging.cpp

```diff
@@ -0,0 +1,303 @@
+                        );
+                        auto uy_new = (
+                            vx * cos_theta * sin_phi + vy * cos_phi
+                            + cluster_u_mag * sin_theta * sin_phi
+                        );
+                        auto uz_new = -vx * sin_theta + cluster_u_mag * cos_theta;
+
+                        // set the previous two particles' attributes according to
+                        // the previous bin's values
+                        const auto part_idx1 = indices[sorted_indices_data[i - 1]];
+                        const auto part_idx2 = indices[sorted_indices_data[i - 2]];
+
+                        w[part_idx1] = total_weight / 2._prt;
+                        w[part_idx2] = total_weight / 2._prt;
+#if !defined(WARPX_DIM_1D_Z)
+                        x[part_idx1] = cluster_x;
+                        x[part_idx2] = cluster_x;
+#endif
+#if defined(WARPX_DIM_3D)
+                        y[part_idx1] = cluster_y;
+                        y[part_idx2] = cluster_y;
+#endif
+                        z[part_idx1] = cluster_z;
+                        z[part_idx2] = cluster_z;
+
+                        ux[part_idx1] = ux_new;
+                        uy[part_idx1] = uy_new;
+                        uz[part_idx1] = uz_new;
+                        ux[part_idx2] = 2._prt * cluster_ux - ux_new;
+                        uy[part_idx2] = 2._prt * cluster_uy - uy_new;
+                        uz[part_idx2] = 2._prt * cluster_uz - uz_new;
+
+                        // set ids of merged particles so they will be removed
+                        for (int j = 2; j < particles_in_bin; j++){
+                            idcpu[indices[sorted_indices_data[i - 1 - j]]] = amrex::ParticleIdCpus::Invalid;
+                        }
+                    }
+                    // restart the tallies
+                    particles_in_bin = 0;
+                    total_weight = 0._prt;
+                    total_energy = 0._prt;
+#if !defined(WARPX_DIM_1D_Z)
+                    cluster_x = 0_prt;
+#endif
+#if defined(WARPX_DIM_3D)
+                    cluster_y = 0_prt;
+#endif
+                    cluster_z = 0_prt;
+                    cluster_ux = 0_prt;
+                    cluster_uy = 0_prt;
+                    cluster_uz = 0_prt;
+                }
+
+                particles_in_bin += 1;
+                const auto part_idx = indices[sorted_indices_data[i]];
+
+#if !defined(WARPX_DIM_1D_Z)
+                cluster_x += w[part_idx]*x[part_idx];
+#endif
+#if defined(WARPX_DIM_3D)
+                cluster_y += w[part_idx]*y[part_idx];
+#endif
+                cluster_z += w[part_idx]*z[part_idx];
+                cluster_ux += w[part_idx]*ux[part_idx];
+                cluster_uy += w[part_idx]*uy[part_idx];
+                cluster_uz += w[part_idx]*uz[part_idx];
+                total_weight += w[part_idx];
+                total_energy += 0.5_prt * mass * w[part_idx] * (
```

Comment:
Does this algorithm only work with non-relativistic energy? If not, you can use the routine `Algorithms::KineticEnergy`.

Reply:
The algorithm could definitely work for relativistic velocities as well, but the current implementation does not account for such corrections. I will switch the energy calculation to use `Algorithms::KineticEnergy` since that calculation will be correct for non-relativistic particles anyway, and it won't produce any more problems in the later parts where relativistic corrections should be added (I think just in the calculation of the product species' velocity magnitude).

Reply:
I now also updated the calculation for the product species' velocity magnitude to include relativistic effects. I was surprised to see this actually made a substantial difference in energy conservation even though I'm testing with 100 eV hydrogen ions.